File: 278.FirstBadVersion/bad.cc
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left = 1, right = n;
        while (left <= right) {
            // int mid = (left + right) / 2;       // 溢出
            int mid = left + (right - left) / 2;
            bool check = isBadVersion(mid);
            if (check) {
                if (mid == 1 || !isBadVersion(mid-1)) {
                    return mid;
                }
                else {
                    right = mid - 1;
                }
            }
            else {
                left = mid + 1;
            }
        }
        return 0;
    }
};

File: 278.FirstBadVersion/bad.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):
#
def isBadVersion(version):
    return False if version < 1702766719 else True;

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        left, right = 1, n
        while left <= right:
            mid = (left + right) / 2
            check = isBadVersion(mid)
            if check:
                if mid == 1 or not isBadVersion(mid-1):
                    return mid
                else:
                    right = mid - 1
            else:
                left = mid + 1

        return 0


if __name__ == '__main__':
    n = 2126753390
    print Solution().firstBadVersion(n)
